x/protorev/keeper/hooks.go

```diff
@@ -0,0 +1,211 @@
+
+// StoreSwap stores a swap to be checked by protorev when attempting backruns.
+func (k Keeper) StoreSwap(ctx sdk.Context, poolId uint64, tokenIn, tokenOut string) {
+	swapToBackrun := types.Trade{
+		Pool:     poolId,
+		TokenIn:  tokenIn,
+		TokenOut: tokenOut,
+	}
+
+	if err := k.AddSwapsToSwapsToBackrun(ctx, []types.Trade{swapToBackrun}); err != nil {
+		ctx.Logger().Error("Protorev error adding swap to backrun from storeSwap", err) // Does not return since logging is last thing in the function
+	}
+}
+
+// GetComparablePoolLiquidity gets the comparable liquidity of a pool by multiplying the amounts of the pool coins.
+func (k Keeper) GetComparablePoolLiquidity(ctx sdk.Context, poolId uint64) (sdk.Int, error) {
+	coins, err := k.poolmanagerKeeper.GetTotalPoolLiquidity(ctx, poolId)
+	if err != nil {
+		return sdk.Int{}, err
+	}
+
+	return coins[0].Amount.Mul(coins[1].Amount), nil
+}
+
+// StoreJoinExitPoolSwaps stores the swaps associated with GAMM join/exit pool messages in the store, depending on if it is a join or exit.
+func (k Keeper) StoreJoinExitPoolSwaps(ctx sdk.Context, sender sdk.AccAddress, poolId uint64, denom string, isJoin bool) {
+	pool, err := k.gammKeeper.GetPoolAndPoke(ctx, poolId)
+	if err != nil {
+		return
+	}
+
+	// Get all the pool coins and iterate to get the denoms that make up the swap
+	coins := pool.GetTotalPoolLiquidity(ctx)
+
+	// Create swaps to backrun being the join coin swapped against all other pool coins
+	for _, coin := range coins {
+		if coin.Denom == denom {
+			continue
+		}
+		// Join messages swap in the denom, exit messages swap out the denom
+		if isJoin {
+			k.StoreSwap(ctx, poolId, denom, coin.Denom)
+		} else {
+			k.StoreSwap(ctx, poolId, coin.Denom, denom)
+		}
+	}
+}
+
+// AfterPoolCreatedWithCoins checks if the new pool should be stored as the highest liquidity pool
+// for any of the base denoms, and stores it if so.
+func (k Keeper) AfterPoolCreatedWithCoins(ctx sdk.Context, poolId uint64) {
+	baseDenoms, err := k.GetAllBaseDenoms(ctx)
+	if err != nil {
+		ctx.Logger().Error("Protorev error getting base denoms in AfterCFMMPoolCreated hook", err)
+		return
+	}
+
+	baseDenomMap := make(map[string]bool)
+	for _, baseDenom := range baseDenoms {
+		baseDenomMap[baseDenom.Denom] = true
+	}
+
+	pool, err := k.poolmanagerKeeper.GetPool(ctx, poolId)
+	if err != nil {
+		ctx.Logger().Error("Protorev error getting pool in AfterCFMMPoolCreated hook", err)
+		return
+	}
+
+	denoms, err := k.poolmanagerKeeper.RouteGetPoolDenoms(ctx, poolId)
+	if err != nil {
+		ctx.Logger().Error("Protorev error getting pool liquidity in afterPoolCreated", err)
+		return
+	}
+
+	// Pool must be active and the number of denoms must be 2
+	if pool.IsActive(ctx) && len(denoms) == 2 {
+		if _, ok := baseDenomMap[denoms[0]]; ok {
+			k.CompareAndStorePool(ctx, poolId, denoms[0], denoms[1])
+		}
+		if _, ok := baseDenomMap[denoms[1]]; ok {
+			k.CompareAndStorePool(ctx, poolId, denoms[1], denoms[0])
+		}
```

Comment:
Please add a comment explaining the logic here in-detail

Reply:
Added here: https://github.com/osmosis-labs/osmosis/pull/5045/commits/8966b6a0e3bf2cc11ba5c13f91f8429a7e014625